keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+
+Some of the feedback we have heard from the community:
+
+- "structs with pointers" feels more go idiomatic and more closely aligned with
+  the go structs used for Kubernetes types both for builtin types and by
+  Kubebuilder.
+- It's nice how "builders" are clearly visually distinct from the main go types.
+- Having to use a utility library to wrap literal values as pointers for the
+  "structs with pointers" is not a big deal. I'm already familiar
+  with having to do this in go and once I learn use a utility library for it
+  I'm all set.
+- The "builders" are awkward to use in an IDE. I felt like I was fighting with
+  my IDE to get chain function calls and organize them hierarchally as expected
+  by this approach.
+
+TODO: We are providing the developer community with a fork of controller-tools
+that will allow them to generate apply configuration types that have both the
+alternatives. Our goal is to get feedback from developers after they try out the
+generated apply configurations and use that to inform our decision.
+
+While it is possible to generate types that have both the pointer fields exposed
+and the builder functions, we would prefer to provide clear guidance to the
+community on how apply configurations should be represented in go and encourage
+consistent use of only one of these approaches.
+
+#### DeepCopy support
+
+If "structs with pointers" approach is used, the existing deepcopy-gen
+can be used to generate deep copy impelemntations for the generated apply
+configuration types.
+
+#### Code Generator Changes
+
+hack/update-codegen.sh and hack/verify-codegen.sh will be updated to generate
+the apply functions and apply configuration types.
+
+##### Addition of applyconfiguration-gen
+
+- Add staging/src/k8s.io/code-generator/cmd/applyconfigurations-gen
+- Generates into staging/vendor/k8s.io/client-go/applyconfigurations/
+- Only generate builders for struct types reachable from the types that have the +clientgen annotation
+- Don't generate builders for MarshalJSON types (Quantity, IntOrString)
+- Don't generate builders for RawExtension or Unknown
+
+##### client-gen changes
+
+Since client-gen is available for use with 3rd party project, we must ensure all
+changes to it are backward compatible. The Apply functions will only be generated
+by client-gen if a optional flag is set.
+
+The Apply functions will be included for all built-in types. Strictly speaking
+this can be considered a breaking change to the generated client interface, but
+adding functions to interfaces is a change we have made in the past, and developers
+that have alternate implementations of the interface will usually get a compiler
+error in this case, which is relatively trivial to resolve
+
+
+### Interoperability with structured and unstructured types
+
+For "structs with pointers", json.Marshal, json.Unmarshal and conversions to and
+from unstructured work the same as with go structs.
```

Comment:
I was thinking about conversion of objects from existing clients/structure to new clients/structure and vice versa. One can always go from one to the other by round-tripping through json/yaml. Would this be useful? Do we have a need for more performance (without going through yaml/json)? Do we want to prevent people from going from existing to new, since that would carry required zero-values? Do we want to implement our own version that doesn't carry zero-value (which would possibly lose some information)?

Reply:
I think the builder should at least have some way of viewing what you've built, e.g. maybe it can emit yaml on demand. Good for debugging and logs.

Reply:
I agree. The existing prototype includes getter functions. I'll mention them in the KEP here (and wherever else it is appropriate) as well.